pkg/controller/certificates/keymanager/keymanager_controller.go

```diff
@@ -303,8 +315,32 @@ func (c *controller) setNextPrivateKeySecretName(ctx context.Context, crt *cmapi
 	}
 	crt = crt.DeepCopy()
 	crt.Status.NextPrivateKeySecretName = name
-	_, err := c.client.CertmanagerV1().Certificates(crt.Namespace).UpdateStatus(ctx, crt, metav1.UpdateOptions{})
-	return err
+	return c.updateOrApplyStatus(ctx, crt)
+}
+
+// updateOrApplyStatus will update the controller status. If the
+// ServerSideApply feature is enabled, the managed fields will instead get
+// applied using the relevant Patch API call.
+func (c *controller) updateOrApplyStatus(ctx context.Context, crt *cmapi.Certificate) error {
+	if utilfeature.DefaultFeatureGate.Enabled(feature.ServerSideApply) {
+		crt := &cmapi.Certificate{
+			TypeMeta:   metav1.TypeMeta{Kind: cmapi.CertificateKind, APIVersion: cmapi.SchemeGroupVersion.Identifier()},
+			ObjectMeta: metav1.ObjectMeta{Namespace: crt.Namespace, Name: crt.Name},
+			Status:     cmapi.CertificateStatus{NextPrivateKeySecretName: crt.Status.NextPrivateKeySecretName},
+		}
+		crtData, err := json.Marshal(crt)
```

Comment:
How does this play with struct fields that are not omitempty? Won't we marshal these fields as empty values? Does this affect ownership?

Reply:
Good question.

Looks like we are OK on the Certificate's since all status fields are omitempty (and we are applying to the status sub-resource). This is the same for type object meta fields too.

We should keep an eye out for this though as we cover the other resources.

Reply:
Can we write tests for this too? e.g. a simple unit test that ensures that the serialised form is as expected? (i.e. calls this function/the serialisation portions of it and compares the string data to 'known good data')?

Then perhaps have an OWNERS file that matches on that specific file in each package and requires an approval from yourself, maybe?

Reply:
See https://github.com/kubernetes/kubernetes/blob/e5ae3f8d67049d60c30b977bda35274ad93667fd/OWNERS#L25 as an example of how that can be done

Reply:
Yep, agreed to all of that :+1: